chainer/optimizer.py

```diff
@@ -61,6 +64,19 @@ def __getattr__(self, name):
             raise AttributeError('_parent is not set up yet')
         return getattr(self._parent, name)
 
+    def __setattr__(self, name, value):
+        # If the attribute is not defined as the class attribute of
+        # `Hyperparameter`, it's assumed to be a hyperparameter.
+        if not hasattr(Hyperparameter, name):
+            if not (isinstance(value, (numpy.ndarray, cuda.ndarray))
```

Comment:
I suppose this change makes Hyperparameter less flexible.

Reply:
Do you mean you may want to set a hyperparameter which is not an array nor a scalar?

Reply:
Yes.
I've encountered the case I'd like to pass a string argument to an optimizer, which syncs parameters `'soft'` or `'hard'`.  To be honest, my final code didn't pass the string, and I've not set a string value to a hyperparameter (because the optimizer was not an instance of `GradientMethod`.)

Reply:
So, we need to decide to which extent `Hyperparameter` accepts as a value.

The most conservative option is to accept anything but `Link`, but in this case we would need to add such rules in the future depending on how it is used. I don't think that's a good choice.

I think the minimum is scalars and ndarrays (as in the current code).
Maybe `str` is reasonable, too.